storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -512,10 +514,146 @@ func (s *peersSource) ReadIndex(
 	shardsTimeRanges result.ShardTimeRanges,
 	opts bootstrap.RunOptions,
 ) (result.IndexBootstrapResult, error) {
-	// FOLLOWUP(r): implement the peers source returning
-	// index segments that are streamed from peers as required.
-	// Try to cache some series metadata on the peers source itself
-	// from work done in ReadData(...) to help avoid rereading as
-	// much as possible.
-	return result.NewIndexBootstrapResult(), nil
+	// FOLLOWUP(r): Try to reuse any metadata fetched during the ReadData(...)
+	// call rather than going to the network again
+	r := result.NewIndexBootstrapResult()
+	if shardsTimeRanges.IsEmpty() {
+		return r, nil
+	}
+
+	session, err := s.opts.AdminClient().DefaultAdminSession()
+	if err != nil {
+		s.log.Errorf("peers bootstrapper cannot get default admin session: %v", err)
+		r.SetUnfulfilled(shardsTimeRanges)
+		return nil, err
+	}
+
+	var (
+		count         = len(shardsTimeRanges)
+		concurrency   = s.opts.DefaultShardConcurrency()
+		dataBlockSize = ns.Options().RetentionOptions().BlockSize()
+		resultOpts    = s.opts.ResultOptions()
+		idxOpts       = ns.Options().IndexOptions()
+		version       = s.opts.FetchBlocksMetadataEndpointVersion()
+		resultLock    = &sync.Mutex{}
+		wg            sync.WaitGroup
+	)
+	s.log.WithFields(
+		xlog.NewField("shards", count),
+		xlog.NewField("concurrency", concurrency),
+	).Infof("peers bootstrapper bootstrapping index for ranges")
+
+	workers := xsync.NewWorkerPool(concurrency)
+	workers.Init()
+	for shard, ranges := range shardsTimeRanges {
+		shard, ranges := shard, ranges
+		wg.Add(1)
+		workers.Go(func() {
+			defer wg.Done()
+
+			iter := ranges.Iter()
+			for iter.Next() {
+				currRange := iter.Value()
+				metadata, err := session.FetchBootstrapBlocksMetadataFromPeers(ns.ID(),
```

Comment:
Sorry I'm just catching this now, but could you take a look at fetchBootstrapBlocksFromPeers? after currRange := iter.Value() it then  steps through the range one block at a time.

This is just metadata so its less important, but you could imagine this causing much higher memory utilization than necessary during node joins in clusters with lots of data and very large disks (I.E you have like 16 Terabytes of storage on a host, and for each shard you're fetching all the IDs/Tags for all of the historical blocks instead of a single block at a time)

Reply:
Also if we decide to do this, a quick check to make sure that the range divides evenly over the blocksize would be good (it always should, but if it doesn't weird things might happen...)

Reply:
Np, will make this the same as the data code path and do it per block.